packages/keyring-controller/src/KeyringController.ts

```diff
@@ -571,34 +610,46 @@ export class KeyringController extends BaseControllerV2<
   async signTypedMessage(
     messageParams: TypedMessageParams,
     version: SignTypedDataVersion,
-  ) {
+  ): Promise<string | Bytes> {
```

Comment:
This `Bytes` return type seems like another type error from `KeyringController`. This should be returning a string at the moment, for all keyrings.

Reply:
This is a very weird case because we use EthKeyringController for signing typed data only for QRKeyring, but as far as I know the QRKeyring has never been used by client through KeyringController.

Indeed if we don't detect a QRKeyring what we do is extract the private key from the keyring and use it with eth-sig-util directly 

The ideal scenario here would be to always use EthKeyringController

Reply:
The way this method works is indeed very strange. We should be passing the message along to the keyring in all cases, not just for the QR keyring. Each keyring has its own implementation of `signTypedData`, we don't need it to be re-implemented here.

In fact, fixing this is a blocker to using this keyring controller for hardware wallets (we can't export accounts on hardware wallets, so this inline implementation won't work).

Reply:
Luckily this type issue was easily fixed: https://github.com/MetaMask/KeyringController/pull/249